# gcl_snp_calling
![](pipeline_dag.png)
## Folder Structure
```
ğŸ“ project-root/
â”œâ”€â”€ ğŸ“ gcl_snp_calling/               # All Nextflow workflow logic
â”‚   â”œâ”€â”€ ğŸ“„ main.nf                    # Entry point for pipeline
â”‚   â”œâ”€â”€ ğŸ“„ nextflow.config            # SLURM and container config
â”‚   â”œâ”€â”€ ğŸ“„ freebayes_parameters.json  # Example Parameter config file for freebayes
â”‚   â”œâ”€â”€ ğŸ“„ README.md                  # Usage and documentation
â”‚   â””â”€â”€ ğŸ“ modules/                   # Individual DSL2 module processes
â”‚       â”œâ”€â”€ create_chunks.nf
â”‚       â”œâ”€â”€ freebayes_chunk.nf
â”‚       â”œâ”€â”€ combine_vcfs.nf
â”œâ”€â”€ ğŸ“ data/bam                       # Input BAM Files
â”œâ”€â”€ ğŸ“ genome/                        # Downloaded and indexed reference
â”œâ”€â”€ ğŸ“ logs/                          # SLURM/Nextflow job logs
```

## To Run

1. Clone repo into your directory.
2. Run the pipeline from the project root:

```bash
nextflow run gcl_snp_calling/main.nf \
    -profile slurm \
    -resume \
    --bams "./data/bam/*.bam" \
    --reference "./genome/genome.fa" \
    --freebayes_config gcl_snp_calling/freebayes_parameters.json \
    --num_chunks 25 \
    --output_dir "data/snps" \
    --output_vcf "all_samples.vcf.gz"
```
3. Or Run in SLURM `sbatch run_snpCall.sbatch {freebayes_parameters.json} 25`